File: src/libs/comm_cpp/src/ice/bumperIceClient.cpp
/*
 *  Copyright (C) 1997-2013 JDE Developers TeamkinectViewer.camRGB
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see http://www.gnu.org/licenses/.
 *
 *  Author : Jose María Cañas <jmplaza@gsyc.es>
			Francisco Miguel Rivas Montero <franciscomiguel.rivas@urjc.es>

 */
#include "jderobot/comm/ice/bumperIceClient.hpp"

namespace Comm {

BumperIceClient::BumperIceClient(Comm::Communicator* jdrc, std::string prefix) {

	this->prefix=prefix;
	
	this->refreshRate=0;
	float fps=jdrc->getConfig().asFloatWithDefault(prefix+".Fps",10);
	this->cycle=(1/fps)*1000000;


	std::string proxy = jdrc->getConfig().asString(prefix+".Proxy");
	Ice::ObjectPrx baseBumper = jdrc->getIceComm()->stringToProxy(proxy);

	if (0==baseBumper){
		this->on = false;
		std::cout << prefix + ".Proxy configuration not specified" <<std::endl;

	}
	else {

		try{
			this->prx = jderobot::BumperPrx::checkedCast(baseBumper);

			if (0 == this->prx){
				this->on = false;
	 	   		std::cout <<"Invalid proxy "+ prefix + ".Proxy" <<std::endl;
	 		}else{
	 			this->on = true;
	 			std::cout << prefix + " connected" << std::endl;
	 		}

		
		}catch (const Ice::ConnectionRefusedException& e) {
			std::cout << prefix +" inactive" << std::endl;
		}
		catch (const Ice::Exception& ex) {
			std::cerr << ex << std::endl;
		}
	}

	this->pauseStatus=false;

}

BumperIceClient::~BumperIceClient() {
	this->on=false;
}


void BumperIceClient::pause(){
	std::cout << "Paused" << std::endl;
	this->pauseStatus=true;
}

void BumperIceClient::resume(){
	std::cout << "resumed" << std::endl;
	this->controlMutex.lock();
		this->pauseStatus=false;
		this->sem.broadcast();
	this->controlMutex.unlock();
}

void BumperIceClient::run(){

	IceUtil::Time last;

	last=IceUtil::Time::now();
	while (this->on){
		if (pauseStatus){
			IceUtil::Mutex::Lock sync(this->controlMutex);
			this->sem.wait(sync);
		}

		try{
			jderobot::BumperDataPtr localBumper=this->prx->getBumperData();

			this->controlMutex.lock();
			this->bumperData.state = localBumper->state;
    		this->bumperData.bumper = localBumper->bumper;
    		this->controlMutex.unlock();

			
		}
		catch(...){
			LOG(WARNING) <<prefix +"error during request (connection error)";
			usleep(5000);

		}


		if ((IceUtil::Time::now().toMicroSeconds() - last.toMicroSeconds()) > this->cycle ){
			DLOG(WARNING) <<prefix + ": Bumper adquisition timeout-";
		}
		else{
			usleep(this->cycle - (IceUtil::Time::now().toMicroSeconds() - last.toMicroSeconds()));
		}
		this->refreshRate=(int)(1000000/(IceUtil::Time::now().toMicroSeconds() - last.toMicroSeconds()));
		last=IceUtil::Time::now();
	}
}

JdeRobotTypes::BumperData  BumperIceClient::getBumperData(){
	JdeRobotTypes::BumperData bumper;
	this->controlMutex.lock();
	bumper = this->bumperData;
	this->controlMutex.unlock();
	return bumper;
}

} /* NS */

File: src/libs/comm_py/comm/ice/bumperIceClient.py
#
#  Copyright (C) 1997-2017 JDE Developers Team
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see http://www.gnu.org/licenses/.
#  Authors :
#       Aitor Martinez Fernandez <aitor.martinez.fernandez@gmail.com>
#

import traceback
import jderobot
import threading
import Ice
from .threadSensor import ThreadSensor
from jderobotTypes import BumperData

class Bumper:

    def __init__(self, jdrc, prefix):
        '''
        Bumper Contructor.
        Exits When it receives a Exception diferent to Ice.ConnectionRefusedException

        @param jdrc: Comm Communicator
        @param prefix: Name of client in config file

        @type jdrc: Comm Communicator
        @type prefix: String
        '''

        self.lock = threading.Lock()
        try:

            ic = jdrc.getIc()
            proxyStr = jdrc.getConfig().getProperty(prefix+".Proxy")
            base = ic.stringToProxy(proxyStr)
            self.proxy = jderobot.BumperPrx.checkedCast(base)

            self.update()

            if not self.proxy:
                print ('Interface ' + prefix + ' not configured')

        except Ice.ConnectionRefusedException:
            print(prefix + ': connection refused')

        except:
            traceback.print_exc()
            exit(-1)

    def update(self):
        if hasattr(self,"proxy") and self.proxy:
            bumper = self.proxy.getBumperData()
            
            self.lock.acquire()
            self.bumper = bumper
            self.lock.release()

    def hasproxy (self):
        '''
        Returns if proxy has ben created or not. 

        @return if proxy has ben created or not (Boolean)

        '''
        return hasattr(self,"proxy") and self.proxy

    def getBumper(self):
        '''
        Returns last Bumper. 

        @return last JdeRobotTypes Bumper saved

        '''
        if self.hasproxy():
            self.lock.acquire()
            bumper = self.bumper
            self.lock.release()
            return bumper

        return None

class BumperIceClient:
    def __init__(self,jdrc,prefix, start = False):
        self.bumper = Bumper(jdrc,prefix)

        self.kill_event = threading.Event()
        self.thread = ThreadSensor(self.bumper, self.kill_event)
        self.thread.daemon = True

        if start:
            self.start()

    # if client is stopped you can not start again, Threading.Thread raised error
    def start(self):
        self.kill_event.clear()
        self.thread.start()

    # if client is stopped you can not start again
    def stop(self):
        self.kill_event.set()

    def getBumperData(self):
        return self.bumper.getBumper()

    def hasproxy (self):
        '''
        Returns if proxy has ben created or not. 

        @return if proxy has ben created or not (Boolean)

        '''
        return self.bumper.hasproxy()
